examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,635 @@
+        )
+        # Compute the loss for the captions using crossentropy
+        captions_loss = keras.losses.categorical_crossentropy(
+            y_true=targets, y_pred=logits, from_logits=True
+        )
+        # Compute the loss for the images using crossentropy
+        images_loss = keras.losses.categorical_crossentropy(
+            y_true=tf.transpose(targets), y_pred=tf.transpose(logits), from_logits=True
+        )
+        loss = (captions_loss + images_loss) / 2
+        # Return the mean of the loss over the batch.
+        return tf.math.reduce_mean(loss)
+
+    def train_step(self, features):
+        with tf.GradientTape() as tape:
+            # Forward pass
+            caption_embeddings, image_embeddings = self(features, training=True)
+            loss = self.compute_loss(caption_embeddings, image_embeddings)
+        # Backward pass
+        gradients = tape.gradient(loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+        # Monitor loss
+        self.loss_tracker.update_state(loss)
+        return {"loss": self.loss_tracker.result()}
+
+    def test_step(self, features):
+        caption_embeddings, image_embeddings = self(features, training=False)
+        loss = self.compute_loss(caption_embeddings, image_embeddings)
+        self.loss_tracker.update_state(loss)
+        return {"loss": self.loss_tracker.result()}
+
+
+"""
+## Train the dual encoder model
+
+In this experiment, we freeze the base encoders for text and image, and make only
+the projection head trainable.
+"""
+
+projection_dims = 256
+num_projection_layers = 1
+dropout_rate = 0.1
+learning_rate = 0.001
+weight_decay = 0.001
+num_epochs = 30
+batch_size = 256
+temperature = 0.05
+
+vision_encoder = create_vision_encoder(
+    num_projection_layers, projection_dims, dropout_rate
+)
+vision_encoder.summary()
+
+text_encoder = create_text_encoder(num_projection_layers, projection_dims, dropout_rate)
+text_encoder.summary()
+
+optimizer = tfa.optimizers.AdamW(learning_rate=learning_rate, weight_decay=weight_decay)
+dual_encoder = DualEncoder(text_encoder, vision_encoder, temperature)
+dual_encoder.compile(optimizer)
+
+"""
+Note that training the model with 60,000 image-caption pairs, with a batch size of 256,
+takes around 12 minutes per epoch using a V100 GPU accelerator. If 2 GPUs are available,
+the epoch takes around 8 minutes.
+"""
+
+print(f"Number of GPUs: {len(tf.config.list_physical_devices('GPU'))}")
+print(f"Number of examples (caption-image pairs): {train_example_count}")
+print(f"Batch size: {batch_size}")
+print(f"Steps per epoch: {int(np.ceil(train_example_count / batch_size))}")
+train_dataset = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+valid_dataset = get_dataset(os.path.join(tfrecords_dir, "valid-*.tfrecord"), batch_size)
+# Create a learning rate scheduler callback.
```

Comment:
Have you tried one of the built-in learning rate schedules? In general I'd recommend using one of these rather than a callback. https://keras.io/api/optimizers/learning_rate_schedules/

Reply:
I came across it as I found that such technique (reducing learning rate if eval quality doesn't improve) is being used in several recent research papers. I am not sure how effective this technique in this particular example, but I think it is cool and I wanted to highlight it :-)